### PR TITLE
[Core] Moving data value container to geometry

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
This is the draft for discussion in #3244. 

Recently we become aware that a major memory occupation of Kratos is due to many copies of DataValueContainer (most of them empty) in all elements and conditions. This was an extra motivation to give it a try and resulted to be less painful than expected. 

It changes the behavior because now the data is stored in geometry and is shared between all elements and conditions sharing that geometry. For example, all elements created by connectivity preserve modeler will share data with their corresponding element sharing the geometry.

### CODE REVIEW

**Comment by loumalouomega** [comment]
I will need to check this in the contact app. I reuse the same geometry for different pairs, but I store some info in the conditions, so may conflict exists

**Comment by loumalouomega** [comment]
Good thing is that the biggest conflict was solved with the pairs of geometries by @tteschemacher

**Comment by RiccardoRossi** [comment]
@loumalouomega please also check the time. as i recall you generate on the flight a lot of conditions so it would be good to know how much is the time impact.

similary @rubenzorrilla please check the build times for one of the large embedded examples. i would like to know if there is a significative tine impact on the buikd (you generate a lot of geometries  when you compute shape functions)

**Comment by loumalouomega** [comment]
> @loumalouomega please also check the time. as i recall you generate on the flight a lot of conditions so it would be good to know how much is the time impact.

Do you mean after the change or now?

**Comment by loumalouomega** [comment]
> Do you mean after the change or now?

OK, now the geometry will be a heavier object and this will have a performance impact

**Comment by loumalouomega** [comment]
I tried in that branch, the increment of time is not significant for the tests (including validation), but even the validation tests are not enormous. For me I think is acceptable this change

File: kratos/includes/condition.h
Comment (tteschemacher) [inline kratos/includes/condition.h]: Shouldn't we keep this and bypass it to the geometry?

**Comment by tteschemacher** [comment]
If the data container is moved to the geometry, will the CL also be enhanced with the data container?

**Comment by RiccardoRossi** [comment]
the CL has access to the geometry, so indirectly it will have access to the DataValue container (cleanly solving #3048)

**Comment by loumalouomega** [comment]
After merging recent changes all tests in the ContactStructural pass, without modification, something that I don't know why... (should not work). Maybe I was cloning already and I am still creating a new dataset each time

**Comment by rubenzorrilla** [comment]
> @loumalouomega please also check the time. as i recall you generate on the flight a lot of conditions so it would be good to know how much is the time impact.
> 
> similary @rubenzorrilla please check the build times for one of the large embedded examples. i would like to know if there is a significative tine impact on the buikd (you generate a lot of geometries when you compute shape functions)

I tried with a 3.5M element case and both the current master and the modifications proposed in this branch give a build time around the 9 seconds. Considering that the common case is that only a few elements of the entire volume mesh require to perform the splitting, these results make perfectly sense.

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee approves the movement of the DataValueContainer to the geometry as it does not have impact on the performance.

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee I fixed last conflicts

**Comment by pooyan-dadvand** [comment]
Again potential flow is failing...

Any clue @philbucher?

**Comment by philbucher** [comment]
The reason it is failing could be actually related to the changes of this PR, but I don't know the details
@marcnunezc @inigolcanalejo do you know what this test is doing?

aside from this the memory error appeared again :roll_eyes:, no idea why
I am compiling locally ...

**Comment by marcnunezc** [comment]
I'm compiling locally as well to check the tests

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:38]: Adjoint elements have a pointer to its primal element, and when we initialize the adjoint elements we copy its data to the primal element that they have stored. I have updated this part using the new methods

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:38]: This was copying what is done in the Structural app, where this operation is performed at a custom process:
https://github.com/KratosMultiphysics/Kratos/blob/f4f9e45e7adbccff85c1ae77da6a63c9279638b4/applications/StructuralMechanicsApplication/custom_response_functions/adjoint_processes/replace_multiple_elements_and_conditions_process.cpp#L79-L81
Maybe this new data container change is also affecting here? @armingeiser @MFusseder

File: applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_processes.cpp
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_processes.cpp]: This test was failing because there was a bug in the test! This means that the reference values of the test were wrong.

r_element was not referencing the Element(1) of the model_part so all the definitons that we do of the element below did not appear in the original model part (specifically, r_element.GetValue(WAKE)=1 was only performed at the copy, not at the original model part element)

With the new changes, both point to the same geometry data (as I understand) so r_element.GetValue(WAKE)=1 was correctly setting the element as WAKE in the original model part.

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/tests/cpp_tests/test_processes.cpp]: @marcnunezc  can you make a separate PR for the fix please?
This way this PR remains clean

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp
Comment (pooyan-dadvand) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:38]: Please note that the flags will work the same. (There are different copies in each condition and element sharing a geometry)

**Comment by pooyan-dadvand** [comment]
@marcnunezc, thanks for adapting the implementation and making the tests run. 

Now I would wait for @MFusseder to give me the green light for merging.

**Review by MFusseder — APPROVED** [review]
I am fine with the proposed changes. It only affects slightly one of the adjoint elements. I will do the changes there after this was merged.

**Comment by pooyan-dadvand** [comment]
Thanks for changes and supports

**Comment by philbucher** [comment]
@KratosMultiphysics/mpm are you aware of this?

**Pull request merged**
